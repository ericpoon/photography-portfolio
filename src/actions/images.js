import firebase from '../firebase/firebase';

const database = firebase.database();

export const setImages = images => ({
  type: 'SET_IMAGES',
  images,
});

export const startDownloadImages = () => {
  return (dispatch) => {
    return database.ref('/images').once('value')
      .then((snapshot) => {
        const images = [];
        snapshot.forEach((childSnapshot) => {
          images.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setImages(images));
      });
  };
};

export const addImage = image => ({
  type: 'ADD_IMAGE',
  image,
});

export const startAddImage = (image) => {
  return (dispatch) => {
    return database.ref('/images').push(image)
      .then((ref) => {
        dispatch(addImage({
          id: ref.key,
          ...image,
        }));
      });
  };
};

export const editImage = (id, updates) => ({
  type: 'EDIT_IMAGE',
  id,
  updates,
});

export const startEditImage = (id, updates) => {
  return (dispatch) => {
    return database.ref(`/images/${id}`).update(updates)
      .then(() => {
        dispatch(editImage(id, updates));
      });
  };
};

export const deleteImage = id => ({
  type: 'DELETE_IMAGE',
  id,
});

export const startDeleteImage = (id) => {
  return (dispatch) => {
    return database.ref(`images/${id}`).remove()
      .then(() => {
        dispatch(deleteImage(id));
      });
  };
};
