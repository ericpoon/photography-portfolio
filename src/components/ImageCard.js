import React from 'react';

class ImageCard extends React.Component {
  constructor(props) {
    super();
    const {
      title = '', subtitle = '', description = '', file = null, fileUrl = '',
    } = props;
    this.state = {
      title, subtitle, description, file, fileUrl,
    };
  }

  onTitleChange = (e) => {
    this.setState({ title: e.target.value.trim() });
  };
  onSubtitleChange = (e) => {
    this.setState({ subtitle: e.target.value.trim() });
  };
  onDescriptionChange = (e) => {
    this.setState({ description: e.target.value.trim() });
  };
  onFileChange = (file) => {
    console.log(file);
    this.setState({ file });
  };
  onSaveClick = () => {
    const {
      title, subtitle, description, file,
    } = this.state;
    const { onSaveClick } = this.props;
    const updates = {
      title, subtitle, description, file,
    };
    if (onSaveClick) onSaveClick(updates);
  };
  onDeleteClick = () => {
    this.setState({ title: '', subtitle: '', description: '' });
    const { onDeleteClick } = this.props;
    if (onDeleteClick) onDeleteClick();
  };
  onDiscardClick = () => {
    const { title = '', subtitle = '', description = '' } = this.props;
    this.setState({ title, subtitle, description });
  };

  isDirty = (field) => {
    const propVal = this.props[field] ? this.props[field] : '';
    const stateVal = this.state[field] ? this.state[field] : '';
    return propVal !== stateVal;
  };

  render() {
    const { onImageClick, editable } = this.props;
    const { title, subtitle, description } = this.state;
    if (editable) {
      return (
        <div className={'image-card'}>
          <div
            className={'image-card__image'}
            style={{ backgroundImage: `url(${this.state.fileUrl})` }}
            onClick={onImageClick}
          >
            <button
              className={'button button--floating-top-left button--danger'}
              onClick={this.onDeleteClick}
            />
          </div>
          <input type={'file'} onChange={e => this.onFileChange(e.target.files[0])} />
          <div>
            <input
              className={
                `image-card__title image-card__title--input${
                  this.isDirty('title') ? ' is-dirty' : ''}`
              }
              value={title}
              onChange={this.onTitleChange}
            />
            <input
              className={
                `image-card__subtitle image-card__subtitle--input${
                  this.isDirty('subtitle') ? ' is-dirty' : ''}`
              }
              value={subtitle}
              onChange={this.onSubtitleChange}
            />
            <textarea
              className={
                `image-card__description image-card__description--textarea${
                  this.isDirty('description') ? ' is-dirty' : ''}`
              }
              value={description}
              onChange={this.onDescriptionChange}
            />
            <div className={'image-card__actions'}>
              <button className={'button'} onClick={this.onSaveClick}>Save</button>
              <button className={'button'} onClick={this.onDiscardClick}>Discard</button>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div className={'image-card'}>
        <div
          className={'image-card__image'}
          style={{ backgroundImage: `url(${this.state.fileUrl})` }}
          onClick={onImageClick}
        />
        <div>
          <h3 className={'image-card__title'}>{title}</h3>
          <h4 className={'image-card__subtitle'}>{subtitle}</h4>
          <p className={'image-card__description'}>{description}</p>
        </div>
      </div>
    );
  }
}

export default ImageCard;
