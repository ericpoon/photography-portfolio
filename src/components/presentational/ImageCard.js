import React from 'react';

const ImageCard = (props) => {
  const {
    title, subtitle, description, fileUrl, onImageClick,
  } = props;

  return (
    <div className={'image-card'}>
      <div
        className={'image-card__image'}
        style={{ backgroundImage: `url(${fileUrl})` }}
        onClick={onImageClick}
      />
      <div>
        <h3 className={'image-card__title'}>{title}</h3>
        <h4 className={'image-card__subtitle'}>{subtitle}</h4>
        <p className={'image-card__description'}>{description}</p>
      </div>
    </div>
  );
};

export default ImageCard;
