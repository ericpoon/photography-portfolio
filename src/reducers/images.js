const initialState = [];

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_IMAGES':
      return action.images;
    case 'ADD_IMAGE':
      return [...state, action.image];
    case 'EDIT_IMAGE':
      return state.map((i) => {
        return i.id === action.id ? { ...i, ...action.updates } : i;
      });
    case 'DELETE_IMAGE':
      return state.filter(i => i.id !== action.id);
    default:
      return state;
  }
};
