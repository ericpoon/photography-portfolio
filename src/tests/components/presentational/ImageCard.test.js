import React from 'react';
import { shallow } from 'enzyme';
import ImageCard from '../../../components/presentational/ImageCard';
import images from '../../fixtures/images';

let wrapper,
  onImageClick;

beforeEach(() => {
  const { title, subtitle, description } = images[0];
  onImageClick = jest.fn();
  wrapper = shallow(
    <ImageCard
      title={title}
      subtitle={subtitle}
      description={description}
      onImageClick={onImageClick}
    />
  );
});

it('should render non-editable ImageCard correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should handle onImageClick', () => {
  wrapper.find('div>div').at(0).prop('onClick')();
  expect(onImageClick).toHaveBeenCalled();
});
