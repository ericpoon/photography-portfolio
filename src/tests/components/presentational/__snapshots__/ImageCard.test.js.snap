// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render non-editable ImageCard correctly 1`] = `
<div
  className="image-card"
>
  <div
    className="image-card__image"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  />
  <div>
    <h3
      className="image-card__title"
    >
      Kowloon
    </h3>
    <h4
      className="image-card__subtitle"
    >
      Dec 31, 2017
    </h4>
    <p
      className="image-card__description"
    >
      Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur
    </p>
  </div>
</div>
`;
